### ISSUE TITLE
Asyncronous environment runners

### ISSUE THREAD

**Issue by gtmanfred** [saltstack/salt#12159]
Here is what I would like to do.

``` yaml
    saltmaster:
      match: 'salt*'
      sls:
        - cloud.gluster

    highstate-gluster:
      match: '*gluster.gtmanfred.com'
      require:
        - saltmaster

    create-cluster:
      match: 'G@gluster:master'
      sls:
        - gluster.master
      require:
        - highstate-gluster

    highstate-mount:
      match: 'mount*'
      require:
        - create-cluster
```

Above is the overstate I run on my salt master to create cloud servers,
install gluster on them, peer them, and then mount them.

``` yaml
    build_profile:
      cloud.profile:
        - profile: ubuntu-8
        - names:
          - one.gluster.gtmanfred.com
          - two.gluster.gtmanfred.com
          - mounter.gtmanfred.com
          - three.gluster.gtmanfred.com
          - four.gluster.gtmanfred.com
          - five.gluster.gtmanfred.com
          - six.gluster.gtmanfred.com
          - seven.gluster.gtmanfred.com
          - eight.gluster.gtmanfred.com
```

This is the sls for /srv/salt/cloud/gluster.sls, the first state run in
the gluster overstate.  Here is the problem, it creates one server at a
time.

What I would really like to be able to do would be to pass something
like salt-run, and then pass my overstate, but for certain things that
are proxy commands, like building the cloud servers, they should be able
to be run on any one of the servers that I already have out there.  Say
I have 4 servers already attached to the salt master, and none of them
are doing anything, they are just named something like
app1-4.gtmanfred.com.  I would love to be able to say

'Hey, everyone that matches app*, first one you build one.gluster,
second you build two.gluster...' etc.

The cool part, if you include salt-api, and we make it so that you can
pass a flag to not wait for the ip to come back and bootstrap the
servers when told to by a webhook through salt-api to the reactor, you
could just have each server send one build command, and then get told oh
and then you also need to build this one.  Even without that, I would be
building one server at a time, where as now, if I did that and didn't
match only one server, I would recieve failures or duplicate servers
based on the servers all building the first one in the list if they are
matched as the target in the overstate.

Basically, I want to be able to distribute proxy tasks that are being
done, instead of just having designate one specific minion as a target
for a proxy task.

**Comment by gtmanfred** [comment]
I believe this has been implemented. using subset.
